text/3268-repr-scalable.md

```diff
@@ -0,0 +1,191 @@
+- Feature Name: repr_scalable
+- Start Date: 2022-05-19
+- RFC PR: [rust-lang/rfcs#3268](https://github.com/rust-lang/rfcs/pull/3268)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Expanding the SIMD functionality to allow for runtime determined vector lengths.
+
+# Motivation
+[motivation]: #motivation
+
+Without some support in the compiler it would be impossible to use the
+[ACLE](https://developer.arm.com/architectures/system-architectures/software-standards/acle)
+[SVE](https://developer.arm.com/documentation/102476/latest/) intrinsics from Arm.
+
+This RFC will focus on the Arm vector extensions, and will use them for all examples. A large amount of what this
+RFC covers is emitting the vscale attribute from LLVM, therefore other scalable vector extensions should work.
+In an LLVM developer meeting it was mentioned that RISC-V would use what's accepted for Arm SVE for their vector extensions.
+\[[see slide 17](https://llvm.org/devmtg/2019-04/slides/TechTalk-Kruppe-Espasa-RISC-V_Vectors_and_LLVM.pdf)\]
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This is mostly an extension to [RFC 1199 SIMD Infrastructure](https://rust-lang.github.io/rfcs/1199-simd-infrastructure.html).
+An understanding of that is expected from the reader of this. In addition to that, a basic understanding of
+[Arm SVE](https://developer.arm.com/documentation/102476/latest/) is assumed.
+
+Existing SIMD types are tagged with a `repr(simd)` and contain an array or multiple fields to represent the size of the
+vector. Scalable vectors have a size known (and constant) at run-time, but unknown at compile time. For this we propose a
+new kind of exotic type, denoted by an additional `repr()`. This additional representation, `scalable`,
+accepts an integer to determine the number of elements per granule. See the definitions in
+[the reference-level explanation](#reference-level-explanation) for more information.
+
+e.g. for a scalable vector f32 type the following could be its representation:
+
+```rust
+#[repr(simd, scalable(4))]
+pub struct svfloat32_t {
+    _ty: [f32],
+}
+```
+`_ty` is purely a type marker, used to get the element type for the LLVM backend.
+
+
+This new class of type has the following properties:
+* Not `Sized`, but it does exist as a value type.
+  * These can be returned from functions.
```

Comment:
This seems to indicate that we need support for "unsized (r)values" to use this feature. Unfortunately the current state of unsized values is "they are a complete mess, and don't even have a consistent MIR-level semantics".

Reply:
We don't currently have support for returning unsized from functions in Rust. I would like for this RFC to better detail the impact that this will have on implementing scalable vectors in Rust. I hope I can provide some helpful information below.

If we look at [this](https://godbolt.org/z/Eesrva6d9) example, we can see that C/C++ can handle:
1. Scalable types as function params
2. Scalable types as local variables
3. Scalable types as return values

## How does C/C++ do it?

These types in C/C++ are both sizeless and scalable sized. It seems that they invoke either of these properties where it is convenient. For example if you try to take `sizeof` on a scalable type:

```
<source>:8:5: error: invalid application of 'sizeof' to sizeless type 'vint32m8_t' (aka '__rvv_int32m8_t')
    8 |     sizeof(vint32m8_t);
```
Another example of the scalable type being sizeless is in [ASTContext::getTypeInfoImpl](https://github.com/llvm/llvm-project/blob/eae7554d3f1fb1546c629a9a2ae0654b1001ab41/clang/lib/AST/ASTContext.cpp#L2164):

```
    // Because the length is only known at runtime, we use a dummy value
    // of 0 for the static length.
#define SVE_VECTOR_TYPE(Name, MangledName, Id, SingletonId, NumEls, ElBits,    \
                        IsSigned, IsFP, IsBF)                                  \
  case BuiltinType::Id:                                                        \
    Width = 0;
```

But on the other hand, clang also treats these types as having a scalable size which can be resolved at runtime. There is a function [getBuiltinVectorTypeInfo](https://github.com/llvm/llvm-project/blob/e71ce0020fa38a8bf0f4f1518e216659b96b4c13/clang/lib/AST/ASTContext.cpp#L3898C13-L3898C37). In this function you can see how a `BuiltinVectorTypeInfo` object gets created using `ElementCount::getScalable`:
```
#define SVE_ELTTY(ELTTY, ELTS, NUMVECTORS)                                     \
  {ELTTY, llvm::ElementCount::getScalable(ELTS), NUMVECTORS};

// ... snip

#define RVV_VECTOR_TYPE_INT(Name, Id, SingletonId, NumEls, ElBits, NF,         \
                            IsSigned)                                          \
  case BuiltinType::Id:                                                        \
    return {getIntTypeForBitwidth(ElBits, IsSigned),                           \
            llvm::ElementCount::getScalable(NumEls), NF};
```

Then in [SemaChecking.cpp](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaChecking.cpp), there are function calls such as `areCompatibleSveTypes`, `checkRVVTypeSupport`, `CheckImplicitConverssion` which type check treating these types as having a scalable size.

When it comes to code-gen to LLVM IR, Rust `unsized` types have been tricky because it can be difficult to lower unsized types, especially when it comes to return types. But that isn't the case with scalable types. Rust scalable types can be mapped to LLVM scalable types. I think this may allow us to sidestep a lot of the complications that come with supporting general `unsized types` in Rust. Using the godbolt example above we see that the C scalable/sizeless types lowered as LLVM scalable types:

```
  %7 = load i64, ptr %4, align 8
  %8 = call <vscale x 16 x i32> @foo(__rvv_int32m8_t, unsigned long)(<vscale x 16 x i32> %6, i64 noundef %7)
  store <vscale x 16 x i32> %8, ptr %5, align 4
```

## Relying on Builtins
One important point I want to make here is that C/C++ is limiting scalable/sizeless types to builtins. For example, you cant define your own scalable type. In addition you cant define data structures using existing builtin scalable types:

```
// This is an error
struct  foo {
vint32m8_t b;
vint32m8_t a;
};
```

As a result, the scope of handling these types is greatly reduced. As I pointed out above, functions like `areCompatibleSveTypes`, `checkRVVTypeSupport` know how to type check specifically on these scalable types. There is explicit lowering of intrinsics that operate on these types. I believe that by restricting support to only care for handling unsized scalable builtins, then we may not have to concern ourselves with what a mess general unsized types are in Rust. 

## What does this mean for Rust

I hope that this RFC can clarify what it will look like to add support for scalable vectors, in the context of `unsized` in Rust. Some questions I would like to clarify:
* Will we support unsized fn params, unsized local variables, and unsized return values in general, or will we limit the scope to scalable types? I am leaning towards the latter, especially because supporting unsized return values might be a massive undertaking, if it is possible at all. I think if you choose the former, then we should have an RFC on adding that feature to the language. I've started inquiring about that topic on [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Unsized.20Return.20Values) in attempt to understand if any work had been done yet.
* Will scalable types be builtin or can people define their own scalable types in their own Rust programs? If we choose the builtin path, I would like this RFC to discuss adding builtins under Prior Art.
* If we sometimes treat these types as `unsized` and sometimes treat them as having scalable sized, what features will we need to include? Would we require something like `#![feature(unsized_fn_params, unsized_locals, unsized_ret_vals)]`, #![feature(scalable_types)]`, or both?